File: python/vayu_sdk/apis/api_customers.py
from functools import singledispatch
from typing import List
from openapi.api.customers_api import CustomersApi
from openapi.models.contact import Contact
from vayu_sdk.clients.vayu_client import VayuClient
from openapi.models.create_customer_request import CreateCustomerRequest
from openapi.models.update_customer_request import UpdateCustomerRequest

from openapi.models.create_customer_response_customer import (
    CreateCustomerResponseCustomer,
)
from openapi.models.list_customers_response import ListCustomersResponse
from openapi.models.get_customer_response import GetCustomerResponse
from openapi.models.create_customer_response import CreateCustomerResponse
from openapi.models.update_customer_response import UpdateCustomerResponse
from openapi.models.delete_customer_response import DeleteCustomerResponse
from openapi.models.address import Address


Customer = CreateCustomerResponseCustomer


class CustomersAPI:
    __client: CustomersApi = None

    def __init__(self, vayu_client: VayuClient):
        vayu_client.validate_logged_in()
        self.__client = CustomersApi(vayu_client.client)

    def list(self, cursor: str = None, limit: int = None):
        return self.__client.list_customers(limit=limit, cursor=cursor)

    def get(self, id: str):
        response = self.__client.get_customer(customer_id=id)

        return response
    
    def getByExternalId(self, external_id: str):
        response = self.__client.get_customer_by_external_id(external_id=external_id)

        return response

    def create(
        self, name: str, external_id: str, aliases: List[str] = [], address: Address = None, contacts: List[Contact] = None
    ):
        request = CreateCustomerRequest(
            name=name, externalId=external_id, aliases=aliases, address=address, contacts=contacts
        )
        response = self.__client.create_customer(create_customer_request=request)
        return response

    def update(self, id: str, name: str = None, external_id: str = None, aliases: List[str] = None, address: Address = None, contacts: List[Contact] = None):
        request = UpdateCustomerRequest(name=name, externalId=external_id, address=address, contacts=contacts, aliases=aliases)
        response = self.__client.update_customer(customer_id=id, update_customer_request=request)

        return response

    def delete(self, id: str):
        response = self.__client.delete_customer(customer_id=id)

        return response

File: go/api/api_customers.go
package api

import (
	"github.com/vayucode/vayu-sdks/go/client"
	"github.com/vayucode/vayu-sdks/go/openapi"
)

type CustomersAPI struct {
	vayuClient *client.VayuClient
}

type Customer = openapi.CreateCustomerResponseCustomer
type Address = openapi.Address
type Contact = openapi.Contact
type IntegrationType = openapi.IntegrationType
type ListCustomersResponse = openapi.ListCustomersResponse
type GetCustomerResponse = openapi.GetCustomerResponse
type GetCustomerByNameResponse = openapi.GetCustomerByNameResponse
type GetCustomerByIntegrationIdResponse = openapi.GetCustomerByIntegrationIdResponse
type CreateCustomerRequest = openapi.CreateCustomerRequest
type CreateCustomerResponse = openapi.CreateCustomerResponse
type UpdateCustomerRequest = openapi.UpdateCustomerRequest
type UpdateCustomerResponse = openapi.UpdateCustomerResponse
type DeleteCustomerResponse = openapi.DeleteCustomerResponse

func NewCustomersAPI(client *client.VayuClient) *CustomersAPI {
	return &CustomersAPI{
		vayuClient: client,
	}
}

func NewAddress(country *string, city *string, addressText *string, state *string, postalCode *string) *Address {
	address := &openapi.Address{}

	if country != nil {
		address.SetCountry(*country)
	}
	if city != nil {
		address.SetCity(*city)
	}
	if addressText != nil {
		address.SetAddressText(*addressText)
	}
	if state != nil {
		address.SetState(*state)
	}
	if postalCode != nil {
		address.SetPostalCode(*postalCode)
	}

	return address
}

func NewContact(email *string, name *string) *Contact {
	return &openapi.Contact{Email: email, Name: name}
}

func NewCreateCustomerRequest(name string, externalId *string, aliases []string, address *Address, contacts []Contact) *CreateCustomerRequest {
	request := &openapi.CreateCustomerRequest{Name: name, ExternalId: externalId, Aliases: aliases, Contacts: contacts}

	if address != nil {
		request.SetAddress(*address)
	}

	return request
}

func NewUpdateCustomerRequest(name *string, externalId *string, aliases []string, address *Address) *UpdateCustomerRequest {
	request := &openapi.UpdateCustomerRequest{Name: name, ExternalId: externalId, Aliases: aliases}

	if address != nil {
		request.SetAddress(*address)
	}

	return request
}

func (api *CustomersAPI) ListCustomers(limit *float32, cursor *string) (*ListCustomersResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.ListCustomers(ctx)

	if limit != nil {
		request = request.Limit(*limit)
	}

	if cursor != nil {
		request = request.Cursor(*cursor)
	}

	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) GetCustomer(customerId string) (*GetCustomerResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.GetCustomer(ctx, customerId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) GetCustomerByExternalId(externalId string) (*GetCustomerResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.GetCustomerByExternalId(ctx, externalId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) CreateCustomer(payload CreateCustomerRequest) (*CreateCustomerResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.CreateCustomer(ctx)
	request = request.CreateCustomerRequest(payload)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) UpdateCustomer(customerId string, payload UpdateCustomerRequest) (*UpdateCustomerResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.UpdateCustomer(ctx, customerId)
	request = request.UpdateCustomerRequest(payload)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) GetCustomerByName(name string) (*GetCustomerByNameResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.GetCustomerByName(ctx, name)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) GetCustomerByIntegrationId(integrationType IntegrationType, integrationId string) (*GetCustomerByIntegrationIdResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.GetCustomerByIntegrationId(ctx, integrationType, integrationId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}

func (api *CustomersAPI) DeleteCustomer(customerId string) (*DeleteCustomerResponse, error) {
	ctx, cancel := client.GenerateContextWithTimeout()
	defer cancel()

	request := api.vayuClient.Client.CustomersAPI.DeleteCustomer(ctx, customerId)
	response, _, err := request.Execute()

	if err != nil {
		return nil, client.BuildVayuErrorFromGenericOpenAPIError(err)
	}

	return response, nil
}
